[WIP] Integration test suite harness for Native Image, feat. Quarkus, Micronaut, Helidon and more (#83)
graalvm/mandrel: integration-tests/README.md

# Native image integration tests
Builds Native image executables for small applications based on some Native image capable runtimes such
as Quarkus, Micronaut and Helidon. It also provides a convenient way of building small test apps for Native image backbox testing.

## Prerequisites

> The TS expects you run Java 11+ and have ```ps``` program available on your Linux/Mac

Reviewer: ```suggestion
The test suite (TS) expects you run Java 11+ and have ```ps``` program available on your Linux/Mac
```

Author: Ack.

and ```wmic``` (by default present) on your Windows system.
Native image builds also require you have the following packages installed:
* glibc-devel
* zlib-devel
* gcc
* libffi-devel

On Fedora/CentOS/RHEL they can be installed with:
```bash
dnf install glibc-devel zlib-devel gcc libffi-devel libstdc++-static
```

On Ubuntu-like systems with:
```bash
apt install gcc zlib1g-dev libffi-dev build-essential
```

## Usage

The basic usage that runs all tests:
```

> mvn clean verify -Ptestsuite

Reviewer: Out of curiosity, why not just `mvn clean verify`?

Author: @zakkak The test apps are modules and they are included in the parent by default, so as you have them imported in your IDE as a one project without any hassle. When you run the test suite, you don't want to build those apps too; you just want to run the TS. The TS builds the apps at its own discretion at the right time.  Hence -Ptestsuite says "Just run the TS, don't build apps/* modules..."

If your Maven-fu could bring a better way of doing this without hampering the IDE experience, I am all ears...

Reviewer: Thanks for the clarification @Karm .

If I understand correctly both `mvn clean verify -Ptestsuite` and `mvn clean verify` will have the same **result**, but `mvn clean verify -Ptestsuite` might be faster in some cases.

The difference is that `mvn clean verify` will decent in the apps and do `mvn clean verify` there as well, while `mvn clean verify -Ptestsuite` will only run in testsuite (it is actually the same with `mvn clean verify -pl testsuite`) and the test suite will implicitly invoke `mvn ... package ...` in the apps to build them as needed.

My Maven-fu ( :) ) approach would probably use `io.quarkus:quarkus-maven-plugin` for creating the native images in the apps pom files, and avoid invoking mvn from the testsuite harness.

The entry would look like this (copied from a quarkus integration test):
```xml
                    <plugin>
                        <groupId>io.quarkus</groupId>
                        <artifactId>quarkus-maven-plugin</artifactId>
                        <executions>
                            <execution>
                                <id>native-image</id>
                                <goals>
                                    <goal>native-image</goal>
                                </goals>
                                <configuration>
                                    <cleanupServer>true</cleanupServer>
                                    <enableHttpUrlHandler>true</enableHttpUrlHandler>
                                    <graalvmHome>${graalvmHome}</graalvmHome>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
```

Not sure it's worth the extra effort though. Just sharing my thoughts.

Author: @zakkak I don't understand how this helps Helidon or Micronaut  (and eventually other modules)...? 
I might be missing the point of what is bothering you in the first place. Is it that the default profile is suited for IDE and not for execution? Why would it be a problem?

Reviewer: No problem here, just discussing and sharing my thoughts, feel free to ignore me :) Sorry for not making this clear.

With this plugin you should be able to create the native-image directly through maven (for any framework), without the need to invoke `mvn` and `native-image` from the test harness. In theory this could make configuration easier as well (e.g. for using a different `native-image` executable.

## Thresholds

The test suite works with ```threshold.properties```, e.g. `./apps/quarkus-full-microprofile/threshold.properties`: 

```
linux.time.to.first.ok.request.threshold.ms=50
linux.RSS.threshold.kB=120000
windows.time.to.first.ok.request.threshold.ms=80
windows.RSS.threshold.kB=120000
```

**THIS IS NOT A PERFORMANCE TEST** The thresholds are in place only as a sanity check to make sure
an update to Native image did not make the application runtime to run way over the expected benevolent values.

The measured values are simply compared to be less or equal to the set threshold. One can overwrite the `threshold.properties`
by using env variables or system properties (in this order). All letter are capitalized and dot is replaced with underscore, e.g.

```
APPS_QUARKUS_FULL_MICROPROFILE_LINUX_TIME_TO_FIRST_OK_REQUEST_THRESHOLD_MS=35 mvn clean verify -Ptestsuite 
```

### Example failures

With a rather harsh threshold of 5ms:

```
org.opentest4j.AssertionFailedError: 
Application QUARKUS_FULL_MICROPROFILE took 27 ms to get the first OK request, 

> which is over 5 ms threshold. ==> expected: <true> but was: <false>

Reviewer: The `expected: <true> but was: <false>` is slightly confusing, I guess it refers to `27 <= 5`?
Unfortunately I can only think about adding the actual check just before `==>`,  to improve this.

Author: @zakkak junit.jupiter.api doesn't give me GreaterThan / LessThan IMHO... So it boils down to this:

```
assertTrue(timeToFirstOKRequest <= timeToFirstOKRequestThresholdMs,
               "Application " + app + " took " + timeToFirstOKRequest
                + " ms to get the first OK request, which is over " +
                timeToFirstOKRequestThresholdMs + " ms threshold.");
```

Reviewer: Thought so. What I am suggesting is:
```java
assertTrue(timeToFirstOKRequest <= timeToFirstOKRequestThresholdMs,
               "Application " + app + " took " + timeToFirstOKRequest
                + " ms to get the first OK request, which is over " +
                timeToFirstOKRequestThresholdMs + " ms threshold. (" + app + " <= " + timeToFirstOKRequest + ")");
```
But that's probably more confusing.

There is a better approach discussed [here](https://stackoverflow.com/a/18766311/1119431) but I am probably nit-picking.

        at org.graalvm.tests.integration.RuntimesSmokeTest.testRuntime(RuntimesSmokeTest.java:131)
        at org.graalvm.tests.integration.RuntimesSmokeTest.quarkusFullMicroProfile(RuntimesSmokeTest.java:147)
```

For logs checking, see an example failure before we whitelisted the particular warning:

```
[ERROR] Failures: 
[ERROR]   RuntimesSmokeTest.micronautHelloWorld:154->testRuntime:91 
  build-and-run.log log should not contain error or warning lines that are not whitelisted.
  See testsuite/target/archived-logs/org.graalvm.tests.integration.RuntimesSmokeTest/micronautHelloWorld/build-and-run.log
  and check these offending lines: 
    [WARNING] Discovered module-info.class. Shading will break its strong encapsulation.
```

> Happy testing!

Reviewer: ```suggestion
**Happy testing!**
```
:)
